[PyTorch Distributed] Move NCCL_DEBUG print to after NCCL init

kwen2501 opened the pull request:
Summary:
NCCL may read env vars from nccl.conf, it would be more accurate to get
envs after NCCL inits

Test Plan: unit test

Differential Revision: D34912492

awgu reviewed (approved):
Approved internally.

kwen2501 committed:
[PyTorch Distributed] Move NCCL_DEBUG print to after NCCL init (#74287)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/74287

NCCL may read env vars from nccl.conf, it would be more accurate to get
envs after NCCL inits

Test Plan: unit test

Reviewed By: ajauhri

Differential Revision: D34912492

fbshipit-source-id: fe43bd66eec7bbec9987f77f68714a48e1309737

pritamdamania87 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -1167,6 +1160,12 @@ std::vector<std::shared_ptr<NCCLComm>>& ProcessGroupNCCL::getNCCLComm(
 
+  // At this point NCCL should have been initialized, hence we can accurately get
+  // the env value even if NCCL sets it by reading from nccl.conf file
+  if (getRank() == 0) {
+    LOG(INFO) << "NCCL_DEBUG: " << parse_env("NCCL_DEBUG");
+  }
```
Shouldn't we print this on all ranks? The log above "ProcessGroupNCCL initialized with following options" is printed on all ranks.

kwen2501 replied:
It is more for verbosity consideration - not sure if a user wants 1000 lines of `NCCL_DEBUG=INFO` printed in the log, for something not that critical. The other options, by contrast, need to be consistent across all ranks, otherwise there would be a hang. So there may still a value in printing them on all ranks.
